File: CS/UnboundDS-Code/Form1.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Windows.Forms;
using DevExpress.Data;

namespace UnboundDS_Code {
    public partial class Form1 : DevExpress.XtraEditors.XtraForm {
        public Form1() {
            InitializeComponent();
            gridView1.OptionsBehavior.AutoPopulateColumns = false;
            CreateDS();
            CreateGridColumns();
            gridControl1.UseEmbeddedNavigator = true;

        }

        public void CreateDS() {
            UnboundSource unboundDS = new UnboundSource();
            UnboundSourceProperty unboundSourceProperty1 = new UnboundSourceProperty() { DisplayName = "ID", Name = "Int", PropertyType = typeof(int) };
            UnboundSourceProperty unboundSourceProperty2 = new UnboundSourceProperty() { DisplayName = "Day of Week", Name = "String", PropertyType = typeof(string) };
            unboundDS.Properties.AddRange(new DevExpress.Data.UnboundSourceProperty[] {
            unboundSourceProperty1,
            unboundSourceProperty2});
            unboundDS.ValueNeeded += unboundDS_ValueNeeded;
            unboundDS.ValuePushed += unboundDS_ValuePushed;
            unboundDS.SetRowCount(100000);
            gridControl1.DataSource = unboundDS;
        }

        void unboundDS_ValuePushed(object sender, UnboundSourceValuePushedEventArgs e) {
            var defaultValue = GetDefaultData(e.RowIndex, e.PropertyName);
            var cellKey = new CellKey(e.RowIndex, e.PropertyName);
            if(object.Equals(defaultValue, e.Value))
                this.Differences.Remove(cellKey);
            else
                this.Differences[cellKey] = e.Value;
        }

        void unboundDS_ValueNeeded(object sender, UnboundSourceValueNeededEventArgs e) {
            if(this.Differences.Count > 0) {
                object rv;
                if(this.Differences.TryGetValue(new CellKey(e.RowIndex, e.PropertyName), out rv)) {
                    e.Value = rv;
                    return;
                }
            }
            e.Value = GetDefaultData(e.RowIndex, e.PropertyName);
        }


        public void CreateGridColumns() {
            gridView1.Columns.AddVisible("Int");
            gridView1.Columns.AddVisible("String");
        }

        private struct CellKey : IEquatable<CellKey> {
            int rowIndex;
            string propertyName;
            public int RowIndex { get { return rowIndex; } }
            public string PropertyName { get { return propertyName; } }
            public CellKey(int rowIndex, string propertyName) {
                this.rowIndex = rowIndex;
                this.propertyName = propertyName;
            }
            public bool Equals(CellKey other) {
                return this.RowIndex == other.RowIndex && this.PropertyName == other.PropertyName;
            }
            public override int GetHashCode() {
                return unchecked(RowIndex * 257 + (string.IsNullOrEmpty(this.PropertyName) ? 0 : this.PropertyName[0]));
            }
            public override bool Equals(object obj) {
                if(obj is CellKey)
                    return Equals((CellKey)obj);
                else
                    return false;
            }
        }

        readonly string[] DefaultStringData = CultureInfo.CurrentCulture.DateTimeFormat.DayNames;
        readonly Dictionary<CellKey, object> Differences = new Dictionary<CellKey, object>();
        object GetDefaultData(int rowIndex, string propertyName) {
            switch(propertyName) {
                case "String":
                    return DefaultStringData[rowIndex % DefaultStringData.Length];
                case "Int":
                    return rowIndex + 1;
                default:
                    return null;
            }
        }
    }
}

File: VB/UnboundDS-Code/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Windows.Forms
Imports DevExpress.Data

Namespace UnboundDS_Code
    Partial Public Class Form1
        Inherits DevExpress.XtraEditors.XtraForm

        Public Sub New()
            InitializeComponent()
            gridView1.OptionsBehavior.AutoPopulateColumns = False
            CreateDS()
            CreateGridColumns()
            gridControl1.UseEmbeddedNavigator = True

        End Sub

        Public Sub CreateDS()
            Dim unboundDS As New UnboundSource()
            Dim unboundSourceProperty1 As New UnboundSourceProperty() With {.DisplayName = "ID", .Name = "Int", .PropertyType = GetType(Integer)}
            Dim unboundSourceProperty2 As New UnboundSourceProperty() With {.DisplayName = "Day of Week", .Name = "String", .PropertyType = GetType(String)}
            unboundDS.Properties.AddRange(New DevExpress.Data.UnboundSourceProperty() { unboundSourceProperty1, unboundSourceProperty2})
            AddHandler unboundDS.ValueNeeded, AddressOf unboundDS_ValueNeeded
            AddHandler unboundDS.ValuePushed, AddressOf unboundDS_ValuePushed
            unboundDS.SetRowCount(100000)
            gridControl1.DataSource = unboundDS
        End Sub

        Private Sub unboundDS_ValuePushed(ByVal sender As Object, ByVal e As UnboundSourceValuePushedEventArgs)
            Dim defaultValue = GetDefaultData(e.RowIndex, e.PropertyName)
            Dim cellKey = New CellKey(e.RowIndex, e.PropertyName)
            If Object.Equals(defaultValue, e.Value) Then
                Me.Differences.Remove(cellKey)
            Else
                Me.Differences(cellKey) = e.Value
            End If
        End Sub

        Private Sub unboundDS_ValueNeeded(ByVal sender As Object, ByVal e As UnboundSourceValueNeededEventArgs)
            If Me.Differences.Count > 0 Then
                Dim rv As Object = Nothing
                If Me.Differences.TryGetValue(New CellKey(e.RowIndex, e.PropertyName), rv) Then
                    e.Value = rv
                    Return
                End If
            End If
            e.Value = GetDefaultData(e.RowIndex, e.PropertyName)
            GetHashCode()
        End Sub


        Public Sub CreateGridColumns()
            gridView1.Columns.AddVisible("Int")
            gridView1.Columns.AddVisible("String")
        End Sub

        Private Structure CellKey
            Implements IEquatable(Of CellKey)
            Private rowIndex_Renamed As Integer
            Private propertyName_Renamed As String
            Public ReadOnly Property RowIndex() As Integer
                Get
                    Return rowIndex_Renamed
                End Get
            End Property
            Public ReadOnly Property PropertyName() As String
                Get
                    Return propertyName_Renamed
                End Get
            End Property
            Public Sub New(ByVal rowIndex As Integer, ByVal propertyName As String)
                Me.rowIndex_Renamed = rowIndex
                Me.propertyName_Renamed = propertyName
            End Sub
            Public Overloads Function Equals(ByVal other As CellKey) As Boolean Implements IEquatable(Of CellKey).Equals
                Return Me.RowIndex = other.RowIndex AndAlso Me.PropertyName = other.PropertyName
            End Function
            Public Overrides Function GetHashCode() As Integer
                Return RowIndex * 257 + Convert.ToInt32((If(String.IsNullOrEmpty(Me.PropertyName), 0, Me.PropertyName.Chars(0))))
            End Function
            Public Overrides Function Equals(ByVal obj As Object) As Boolean
                If TypeOf obj Is CellKey Then
                    Return Equals(DirectCast(obj, CellKey))
                Else
                    Return False
                End If
            End Function
        End Structure

        Private ReadOnly DefaultStringData() As String = CultureInfo.CurrentCulture.DateTimeFormat.DayNames
        Private ReadOnly Differences As New Dictionary(Of CellKey, Object)()
        Private Function GetDefaultData(ByVal rowIndex As Integer, ByVal propertyName As String) As Object
            Select Case propertyName
                Case "String"
                    Return DefaultStringData(rowIndex Mod DefaultStringData.Length)
                Case "Int"
                    Return rowIndex + 1
                Case Else
                    Return Nothing
            End Select
        End Function
    End Class
End Namespace
